### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by dfed**
After a few weeks‘ of work, we have a working repository, and now: a readme!

### CODE REVIEW

**Review by ahmdmhasn — APPROVED** [review]

File: README.md
Comment (kierajmumick) [inline README.md:216]: I think we probably should provide a matrix illustrating the situations when one should use instantiated, forwarded, and received. It would make it clear at a glance what to use in what scenarios

**Review by kierajmumick — COMMENTED** [review]
All looks pretty good! I have some comments and suggestions!

File: README.md
Comment (ahmdmhasn) [inline README.md:44]: The inclusion of graphs enhances clarity regarding the various types. Without prior familiarity with the potency of dependency injection in the tree, comprehending this concept initially might prove challenging.

Comment (ahmdmhasn) [reply README.md:44]: Also, I would suggest to mention the different Macros ahead and then jump into an example of each.

Comment (ahmdmhasn) [reply README.md:44]: The intro is fine in this part

File: README.md
Comment (ahmdmhasn) [inline README.md:121]: I think you should keep the same order as in the intro, so `@ExternalInstantiable` should come after `@Received`

File: README.md
Comment (ahmdmhasn) [inline README.md:151]: Is it possible for Macros to deduce the identifier from the type name? Since most protocols typically have a sole default concrete implementation, adopting this approach would significantly enhance the readability of the API. 

Also, to facilitate understanding for beginners, I would recommend extending the earlier comment by simplifying the example. Specifically, suggesting a single method without parameters, such as public static func instantiate() -> Type, would make its clarity and purpose evident to the reader.

File: README.md
Comment (ahmdmhasn) [inline README.md:178]:
```suggestion
Property declarations within [`@Instantiable`](#instantiable) types decorated with the [`@Instantiated`](Sources/SafeDI/PropertyDecoration/Instantiated.swift) are instantiated when its enclosing type is instantiated. `@Instantiated`-decorated properties are available to be [`@Received`](#received) by objects instantiated further down the dependency tree.
```
It has been defined earlier as a Macro, no need for repeating.

File: README.md
Comment (ahmdmhasn) [inline README.md:214]:
```suggestion
Property declarations within [`@Instantiable`](#instantiable) types decorated with the [`@Forwarded`](Sources/SafeDI/PropertyDecoration/Forwarded.swift) are forwarded into the SafeDI dependency tree by a [`ForwardingInstantiator`](Sources/SafeDI/DelayedInstantiation/ForwardingInstantiator.swift) instance’s `instantiate(…)` method. A `@Forwarded`-decorated property is available to be [`@Received`](#received) by objects instantiated further down the dependency tree.
```

File: README.md
Comment (ahmdmhasn) [inline README.md:220]:
```suggestion
Property declarations within [`@Instantiable`](#instantiable) types decorated with the [`@Received`](Sources/SafeDI/PropertyDecoration/Received.swift) are injected into the enclosing type‘s initializer. Received properties must be [`@Instantiated`](#instantiated) or [`@Forwarded`](#forwarded) by an object higher up in the dependency tree.
```

File: README.md
Comment (cyrilchandelier) [inline README.md:308]: > When utilizing SafeDI, you must rely on `@Instantiated`-decorated properties to instantiate your dependencies for you.
What happens if you don't? Say you start passing around a `@Instantiated` dependency your class received, will this cause any issue?

File: README.md
Comment (ahmdmhasn) [inline README.md:302]: To highlight its strength, a useful method is to compare it with other Dependency Injection (DI) frameworks. Presenting the same key points in a table format, using X and ✔️, can effectively illustrate this comparison.

File: README.md
Comment (cyrilchandelier) [inline README.md:85]: Any chance this initializer too can be generated?

File: README.md
Comment (ahmdmhasn) [inline README.md:256]: Can we add another simple example in UIKit for an App level injection as well?

File: README.md
Comment (dfed) [reply README.md:40]: fair. I reached for semicolons because these clauses are _long_ but... that's not a great reason 😅. I'll clean this up.

File: README.md
Comment (dfed) [reply README.md:72]: solid insight. could have one example without a protocol and another subsection+example with a protocol. starting with the simpler example makes sense.

File: README.md
Comment (dfed) [reply README.md:145]: I hadn't thought about combining them! The fact that the implementations/capabilities are quite different doesn't really matter much from a "do we need a new macro" point of view. Certainly doable to combine. I'll give that a shot. Fewer macros would likely make the library simpler to understand.

File: README.md
Comment (dfed) [reply README.md:216]: Extremely smart. We're in need of better visual organization of this README for sure, and a table would be a solid step forward here.

File: README.md
Comment (dfed) [reply README.md:252]: Yes! If you have a `@Forwarded` property you _must_ use a `ForwardingInstantiator` to instantiate the thing. Sounds like I need to make that more clear.

File: README.md
Comment (dfed) [reply README.md:121]: Per Kieraj above I'm likely to collapse `@ExternalInstantiable` into `@Instantiable` which will simplify the ordering a bit.

File: README.md
Comment (dfed) [reply README.md:151]: Identifiers are a Valet-specific thing... I should probably use the `UserDefaults` example from the example project rather than inventing a newer, more complicated example here 😅. Good thinking that this is a bit more complex than it needed to be.

File: README.md
Comment (dfed) [reply README.md:302]: I have a `Comparing SafeDI to other DI libraries` section, but it's pretty light-weight. I like the idea of a table here, but I'm not sure we need to go that hard just yet. This section is more of an adoption guide.

File: README.md
Comment (dfed) [reply README.md:308]: You can call `init` yourself, but then you're basically exiting the SafeDI dependency tree and there are no guarantees re when something is instantiated. Sounds like I should make that more clear here 🙂

File: README.md
Comment (dfed) [reply README.md:85]: In the simple case, it can! I've got [a TODO in the codebase to simplify this](https://github.com/dfed/SafeDI/blob/76b713a766232f203f72e1eeb303d01a81d69cfc/Sources/SafeDIMacros/Macros/InstantiableMacro.swift#L83).

When you have non-injected properties that must be set in `init`, though, you'll still need to write your own initializer. I like that today we can generate the structure of the initializer for you with a FixIt – if I started generating the initializer for you we'd lose the FixIt. I haven't figured out if the trade-off of simplifying the simple case while making the more complex (but still common!) case a bit more annoying is worth it yet.

Of course, now that I've typed that out... I could generate the FixIt only if you have a local property with an initializer... that'd be the best of both worlds. We’ll Get There™. Just gotta put in the time 🙂

File: README.md
Comment (kierajmumick) [reply README.md:252]: Could Forwarded be a property wrapper (instead of a macro) that automatically wraps the forwarded type in a `ForwardingInstantiator`?

File: README.md
Comment (kierajmumick) [inline README.md:77]: I know we talked about this probably about 100 times, and I know you disagree due to differences in coding style across the industry, but seeing this in front of me really does make me want to say this again 😂 

I really do think we should be able to infer that `XYZProtocolName: ProtocolName` fulfills the type `ProtocolName`, _unless_ `fulfillingAdditionalTypes` is provided as an argument.

Otherwise we're writing effectively boilerplate (We're writing the string `UserService` 3 times), which feels like something Macros should eliminate.

File: README.md
Comment (dfed) [reply README.md:256]: We can, but honestly I have significantly less-strong opinions about the right way to define a root using SafeDI in UIKit-world. 

Because `AppDelegate` and `SceneDelegate` must be separate objects, and neither object has a good way to reference the other (outside of the scene delegate referencing and casting `UIApplication.shared.delegate`)... there's no _great_ way to make an application root like there is in `SwiftUI`.

In my own spikes I've been resorting to making a
```
@Instantiable
public final class Root {

    public init(...) { ... }

    static let shared = Root() // uses initializer generated by SafeDI

    // Instantiated properties here
    ...
}
```

And then utilizing `Root.shared` in both the `AppDelegate` and `SceneDelegate`. This is a pretty kludgy (but workable) approach, and is similar to how folk tend to share objects between these two types anyways. Alternatively, you could have separate dependency trees for your `AppDelegate` and `SceneDelegate`... but that seems like a recipe for pain since scenes and the app do need to share context.

So... yeah. I didn't want to have an opinion on this in the README, since there is no great option. Does that make sense?

I agree that having a UIKit example will aid folk who are adopting this system in UIKit, but also... it's not hard to make the leap from "how do I share dependencies between these two delegates" to "I should create a self-vending singleton".

File: README.md
Comment (dfed) [reply README.md:252]: sorry, the `ForwardingInstantiator` is how you instantiate the _enclosing type_ that has a forwarded property, not the forwarded property itself. No way to really make this magical, especially because the `ForwardingInstantiator` needs the `@Forwarded` property as an argument in its `instantiate(...)` method.

Comment (kierajmumick) [reply README.md:252]: from what I remember, the macro is only used as a syntax flag for the plugin, so making this a property wrapper should work.

Comment (kierajmumick) [reply README.md:252]: if i'm understanding this right

File: README.md
Comment (dfed) [reply README.md:77]: We disagree on the point of a macro! The point isn't eliminating boilerplate as much as making it easier to do a standard thing.

At a high level: DI systems are already quite magical, so I really want to avoid "we made this assumption, didn't have a good way to tell you that we did so, and gave you no way to opt out without changing your type name" scenarios.

Imagine we had:
```
public final class ViewController: UIViewController { ... }

@Instantiable
public final class ProfileViewController: ViewController { ... }
```

I kinda doubt we would want to make `ProfileViewController` also fulfill `ViewController`. And if `ViewController` were `@Instantiable` the decision tree gets even more complex. Of course, we could come up with some set of rules to handle these cases and document them here, but... this kind of magic would make the learning curve steeper for this library.

The learning curve for adopting automated DI is already kinda reasonably high. I do not want to make it more complicated when the alternative is "repeat yourself in a way that makes it very clear what you're doing".

File: README.md
Comment (MrAdamBoyd) [reply README.md:308]: If someone forgets to add `@Instantiated` to a property, is this something they will notice easily?

File: README.md
Comment (dfed) [reply README.md:252]: oh yeah I've played with utilizing property wrappers instead of macros and it does work (I did this – and then deleted – for `@LazyInstantiated`).

Are you saying you would have something like
```
@Instantiated // macro
@Forwarded<ForwardArgument> // property wrapper
let builder: Instantiator<InstantiatedType>
```

Because we can't make `@Forwarded` itself a property wrapper that exposes a `ForwardingInstantiator` because the forwarding instantiator is the thing we need to _create the object with the forwarded property_.

All that said: what problem are we solving here?

File: README.md
Comment (dfed) [reply README.md:308]: short answer: yes they will notice quickly and easily.

If they fail add `@Instantiated`, `@Received`, or `@Forwarded` to a property on an `@Instantiatable` type, then SafeDI will fail the build (with a FixIt) until they write an `init` method that does not include this property as an argument. If they accept the FixIt, they'll need to instantiate the property themselves in `init(...)`, which should be a sign that they're not using DI for this property.

File: README.md
Comment (kierajmumick) [reply README.md:252]: okay I misunderstood—I thought you were saying this needed to be:
```
let builder: FordwardingInstantiator<InstantiatedType>
```

But considering that it is not, property wrappers don't make sense here. 

I was less trying to solve a problem, and more trying to understand how this works.

Comment (dfed) [reply README.md:252]: Got it. We need:
```
let builder: FordwardingInstantiator<ForwardedArgument, InstantiatedType>
```

The code example a few lines down hopefully makes this a bit more clear... but I'm very much hearing I need more English explanation of this requirement too.

File: README.md
Comment (dfed) [reply README.md:178]: 😅 can you tell which sections I wrote first? yeah I can simplify here.

File: README.md
Comment (dfed) [reply README.md:145]: Just took a pass at unifying in 0574004a5e3901707ad00790e5cac03644b8c7d8. Still need to update the docs here.

Comment (dfed) [reply README.md]: Docs have been updated!

File: README.md
Comment (dfed) [reply README.md]: Dramatically simplified the README examples

File: README.md
Comment (dfed) [reply README.md]: Added a "macro cheat sheet". The table in this section copied wording from elsewhere but I think the table representation is still helpful as an easy reference. LMK what you think.

File: README.md
Comment (dfed) [reply README.md:243]: I updated the `@Forwarded` description to make this a bit more clear.

File: README.md
Comment (dfed) [reply README.md:48]: See the new cheat sheet section, and the updated "Using SafeDI" section. I think I've addressed what you're getting at. Let me know!

File: README.md
Comment (dfed) [reply README.md]: I've now answered Cyril's q in this section!

File: README.md
Comment (kierajmumick) [inline README.md:72]: I think this is missing an `@Instantiable` macro 😂

File: README.md
Comment (kierajmumick) [inline README.md:131]: This is fufilling **additional** types because `DefaultUserService` will be used to fulfill dependencies of type `UserService` _and_ `DefaultUserService`. Is that right? Asking to make sure I understand clearly.

File: README.md
Comment (kierajmumick) [inline README.md:155]: Nice!! this looks great.

File: README.md
Comment (dfed) [reply README.md:72]: Oh my god. Simplified this example a little too much didn't I.

Comment (dfed) [reply README.md:72]:
```suggestion
@Instantiable
public final class UserService {
```

File: README.md
Comment (dfed) [reply README.md:132]: Correct!

**Review by kierajmumick — APPROVED** [review]
Looks great to me!

File: README.md
Comment (dfed) [reply README.md:51]: alright undid the cheat sheet sectioned put in a table at the top.

**Comment by dfed** [comment]
Going to leave this PR open for a few more days to see if I can get another review pass with the new changes. Once this gets merged I'll release a 0.1.0 beta version of SafeDI!

File: README.md
Comment (NickEntin) [inline README.md:293]: With the exception of the root, right?

File: README.md
Comment (NickEntin) [inline README.md:313]:
```suggestion
It is strongly recommended that projects adopting SafeDI start their migration by identifying the root of their dependency tree and making it `@Instantiable`. Once your root object has adopted SafeDI, continue migrating dependencies to SafeDI in either a breadth-first or depth-first manner. As your adoption of SafeDI progresses, you‘ll find that you are removing more code than you are adding: many of your dependencies are likely being passed through intermediary objects that do not utilize the dependency except to instantiate a dependency deeper in the tree. Once types further down the dependency tree have adopted SafeDI, you will be able to avoid receiving dependencies in intermediary types.
```

File: README.md
Comment (dfed) [reply README.md:63]: They cannot. There must be one initializer that has arguments only for only injected (Instantiable, Forwarded, or Received) properties.

Under the hood, the reason for this is that this is the initializer that SafeDI uses to initialize your Instantiable object. And since SafeDI is only aware of (and has access to) properties in the DI tree... we wouldn't be able to make a call to an initializer with additional properties.

File: README.md
Comment (dfed) [reply README.md:293]: Good call that I should make this explicit. Yeah, for the root of the dependency graph you must call `init()`. It's still true that you shouldn't call `init(...)` though... but yeah I should make this more clear.

File: README.md
Comment (dfed) [reply README.md:63]: Going to leave this as-is for now since:

1. I'm struggling to come up with better wording here
2. There is a FixIt if you do this wrong

File: README.md
Comment (dfed) [reply README.md]: 521a99a4c308e9bceef1065372647cb75fe85266

File: README.md
Comment (dfed) [reply README.md:63]: I need to take another pass here given 418bada – this initializer is now generated for you unless you have uninitialized properties on the type.

File: README.md
Comment (gonzalonunez) [inline README.md:13]: Aesthetic need, I would maybe consider using regular markdown checkboxes here. If anything just to save you some space and get more content above-the-fold:

- [x] Compile-time safe
- [x] Thread safe
- [x] ...

File: README.md
Comment (gonzalonunez) [inline README.md:36]: A single top-level code sample, maybe even just a condensed version of the example app's root view, would be really helpful to see before diving into the specifics behind each macro. Would really help contextualize the details below.

File: README.md
Comment (gonzalonunez) [inline README.md:363]: This feels like an important subtlety. Maybe a code-sample or a link to a Wiki in the repo that explains in detail?

Comment (dfed) [reply README.md:363]: :100:. Will need to do this on a future pass, but yeah this is a tough one. The example app shows has an example of this in practice at least.

File: README.md
Comment (dfed) [reply README.md:36]: Yeah I need to do another pass on this and pull a bunch of the below out into a wiki + focus the README on "how do I get started". That'll be a future round though.

File: README.md
Comment (kierajmumick) [reply README.md:363]: While it’s important and subtle, it is a subtlety inherent in SwiftUIs architecture, with or without this DI system. 

It shouldn’t be the job of these docs to
the SwiftUI subtleties in detail. I do think the call out & link is likely enough!

**Comment by dfed** [comment]
I'm going to merge this, but I am still quite open to feedback on both this documentation and the API! I'll be pushing a beta 0.1.0 tag release of this repo shortly, but since we're still in beta it's a good time to provide API feedback!

One bit of feedback I've received offline from a couple folk is that much of the content in this README would be best split out into wiki docs, and the README should be refocused on more of a Quickstart guide (with clear code examples pulled from the sample project) + linking to the rest of the wiki. This is great feedback, and I agree with it, but in the spirit of shipping I'm going to merge this as-is and take a future action item to rework this.

**Pull request merged**
